### PR TITLE
Dynamic Editor-Buttons TinyMCE

### PR DESCRIPTION
**Pull request by alimpam**
Pull Request for Issue # .

### Summary of Changes
- Make TinyMCE-Plugins (Buttons) configurable (enable / disable, accessrights etc..)
- Makes easier to extend TinyMCE
- easier to maintenance
vorhernachherdragdrop

### Testing Instructions
Before:
You cant enable / disable / install / uninstall TinyMCE-Buttons with Joomla
After:
I created an example for TinyMCE-Smileys-Plugin
1. Download https://github.com/alimpam/tinymce-smileys-plugin and install it
2. In the Plugin-Manager you will see a new Plugin **Smileys-Plugin for TinyMCE** (belongs to editors_buttons category) which can be enabled / disabled
3. You can edit an article and checkout if it works. So if it is enabled you can insert smileys ;-)

### Documentation Changes Required

This PR makes it easier to expand the functionality of TinyMCE and
allows to install/uninstall (as Joomla-Plugins) or enable/disable
Buttons. For that there is a new plugin category "editors_buttons" which
must be used. Also jdragdrop is fitted for this changes. (based on #11157)

### CODE REVIEW

File: plugins/editors/tinymce/tinymce.php [hunk 1]
```diff
@@ -525,471 +449,300 @@ public function onDisplay($name, $content, $width, $height, $col, $row, $buttons
...
 			case 0: /* Simple mode*/
-				$script .= "
-			menubar: false,
-			toolbar1: \"bold italics underline strikethrough | undo redo | bullist numlist | code | $toolbar5\",
-			plugins: \"$dragDropPlg code\",
-		});
-		";
+				$scriptOptions['menubar']  = false;
+				$scriptOptions['toolbar1'] = 'bold italics underline strikethrough | undo redo | bullist numlist | code | ' . $toolbar5;
+				$scriptOptions['plugins']  = 'code ' . implode(',', $extraButtons);
```
Comment (dgt41) [inline plugins/editors/tinymce/tinymce.php]: @alimpam after code you use space and then in you are imploding with comma. If space works then I propose to use it also in the implode e.g. `implode(' ', $extraButtons);`
Nice work! 👍

**Comment by RonakParmar** [comment]
Can you please provide steps to test this PR?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11829">issues.joomla.org/joomla-cms/11829</a>.</sub>

**Comment by schnuti** [comment]
Beside the test instructions. Please explain what kind of functionality it adds that you can't do as is.
Did you base your code on #11157? I think a lot of code is already in that PR.

Personally I like any functionality/usability added to the editor. :+1:

**Comment by brianteeman** [comment]
Now that you have created a real plugin for jdragdrop wouldn't it be better if the associated javascript was moved with it

**Comment by dgt41** [comment]
@brianteeman but it is:
```xml
    <media folder="jdragdrop" destination="editors/tinymce/plugins/jdragdrop">
         <file>plugin.js</file>
         <file>plugin.min.js</file>
     </media>
```

**Comment by brianteeman** [comment]
I meant outside of tinymce own root. It will make maintenance of tinymce
much easier. But maybe that's just me?

On 29 Aug 2016 9:04 p.m., "Dimitri Grammatikogianni" <
notifications@github.com> wrote:

@brianteeman <https://github.com/brianteeman> but it is:

    <media folder="jdragdrop" destination="editors/tinymce/plugins/jdragdrop">
         <file>plugin.js</file>
         <file>plugin.min.js</file>
     </media>

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/joomla/joomla-cms/pull/11829#issuecomment-243238512>,
or mute the thread

**Comment by alimpam** [comment]
@schnuti @RonakParmar - now there are testing instructions

**Comment by dgt41** [comment]
@brianteeman there was a decision that tinymce plugins will be stored in the location `editors/tinymce/plugins` as it makes the integration way to easier to implement. The maintenance is a chapter which needs to be rewritten in j4 with npm/gulp (or similar tools)

**Comment by brianteeman** [comment]
Just seems wrong to me.

**Comment by schnuti** [comment]
I agree with @brianteeman 
1. Why not use the in TinyMCE built in functionality to load scripts from "external sources"? I tested it as in #11157.  'path/to/plugin.js' can point to e.g. 
JUri::root().'media/cms/js/tinymce/cmsbutton.min.js'
or 
JUri::root().'media/com_mycomp/js/tinymce/mybutton.min.js'
or
real external path like a CDN.
You could define rules to allow overrides.
2. How to activate other (new) standard TinyMCE plugins and buttons without coding?

**Comment by schnuti** [comment]
Another thing is the buttons. Wouldn't it be better to merge all (6) cms buttons into one plugin.js?
This only means that the buttons have to be defined in another way then by loaded xtd-plugin.  "button-Article" instead of "buttonX-Article". May not be possible in 3.x?

**Comment by schnuti** [comment]
Sorry for not keeping this in one post.

The new plugin type should return a second parameter with the button(s). In that way one plugin can define more than one button or no buttons at all. Plugin code is not necessary = one button. 
If you need more dynamic buttons the code has to be splitted.
Why not also allow to return also without plugin but only buttons or more then one plugin?

array(array('the plugins'), array('the buttons')

**Comment by mbabker** [comment]
Is this still being worked on?

**Comment by schnuti** [comment]
I hope not. The tinyMCE "workflow" has changed a lot since  aug. 2016. Any proposal would do better in a new issue.

**Comment by franz-wohlkoenig** [comment]
closed as comments above. Feel free to reopen if needed.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11829">issues.joomla.org/tracker/joomla-cms/11829</a>.</sub>

**Comment by dgrammatiko** [comment]
@franz-wohlkoenig can you please reopen this one?

**Comment by ghost** [comment]
done

**Comment by brianteeman** [comment]
@dgrammatiko one year ago you asked for this abandoned pr to be re-opened as there has been no further activity I am closing it again as abandoned. If you wanted it to be kept open as a todo-list reminder for you then I guess it didn't work - and this isnt the place

**Pull request closed without merging**